### DISCUSSION TITLE
Adding support to play theme media in user specified order

### DISCUSSION THREAD

**Discussion by ItsAllAboutTheCode** [jellyfin/jellyfin-web#5958] [Ideas]
I am looking to discuss a feature request to be able add support to the Jellyfin web client to be able to select the order that theme media(music and videos) play when navigating a show or season. The link to that feature request is located at https://features.jellyfin.org/posts/2740/select-audio-from-the-theme-music-folder-at-random.

Jellyfin has support for playing theme media for a Series or Season folder as explained [here](https://jellyfin.org/docs/general/server/media/shows/#other). For theme videos it supports adding files located in a `backdrop/` folder. into a play queue. For theme music it combines any songs that that are named `theme.*` as well as songs that are in a `theme-music/` folder into an audio play queue.

Currently Jellyfin adds any theme-media in sort name order to the play queue. There is no way to alter the order when navigating a show.

I have a Pull Request out that allows the user to select the sort order in which they would like to add the theme media to the play queue at https://github.com/jellyfin/jellyfin-web/pull/5714.

The UI for it is that the option to select the sort order for the theme-media is hidden until the user either selects display options on whether they want to play theme-music or theme-videos.

https://github.com/user-attachments/assets/afd60878-2f94-4689-aeab-f17dc7b53f41

* My questions are is this a feature that the community would want?

* Also is the UI design for it the correct way to go about it? 
1. Hide the UI  to select the order to add theme-media to the play queue until the user selects the option, that they want to play music or videos?
    Or always show the UI for the for selecting the theme-media sort order?
2. Is the wording for the new UI options clear? Should it mention "the order to add theme media to the Play Order Queue", instead of mentioning terms like "Sort Order" or "Sort By"?

**Comment by ItsAllAboutTheCode** [comment]
To illustrate how the feature theme media order can be adjusted according to the Sort order options I have attached an image of a Show with a `Season 01` folder, containing both a `theme-music` and a `theme.mp3` file.
As well as video showing the play queue order when selecting the different sort options in the Jellyfin client.
![theme-music](https://github.com/user-attachments/assets/8afd76e7-a720-4533-9615-7d0a1fc0e549)

https://github.com/user-attachments/assets/b949bb7f-eab5-402a-b856-7e240c14239e

**Comment by erikbucik** [comment]
I would advise against implementing this UI, it's very confusing.

**All these options should be placed into a modal container**, accessible by clicking one list button.

- The modal's first option should be a switch to enable the playback of media in the background while browsing, with two nested checkboxes for including music and videos. _Or something similar._

- The next option would be the "Sort by" dropdown (with no description as the option is self-explanatory).
- And for the ascending/descending order, there are much better components than another dropdown for just two options.


Deciding on what the strings should say is a separate issue (UX Copywriting).
- "Background playback" - that's not clear enough...
_I'm thinking whether the description for the list item is even necessary..._
- "Play theme media in the background"___________[off/on] _or a checkbox_
  - [ ] songs/music
  - [ ] videos
- "Sort queue by"
  - ascending/descending - could just use an arrow icon if it's clear enough?

_^ More thought should definitely be put into this. + some mockups & prototypes drawn._

**Reply by ItsAllAboutTheCode** [reply]
That makes sense how the UI can confuse users as it just "pops up" with no context.

I can prototype the changes to the UI to have a modal that contains the background playback options.

For the modal container, do you know of any reference that can be used as an example?
Do you envisions having a list button like the one in the Settings page ?
![2024-08-25_18-52](https://github.com/user-attachments/assets/43f0fafd-9563-4468-9913-2a119f9daf88).

But instead have the button be on the "Display Settings"  page?
Would it be its own web page?

I have some questions around the 'Sort queue by" option. 
Do you know if Jellyfin has arrow icons already that are available for use? 
As an alternative the [unicode up and down arrows](https://wikiless.ditatompel.com/wiki/Arrow_(symbol)#Unicode) or two radio buttons could be used to show the difference between ascending/descending?

**Reply by ItsAllAboutTheCode** [reply]
> I have some questions around the 'Sort queue by" option. Do you know if Jellyfin has arrow icons already that are available for use? As an alternative the [unicode up and down arrows](https://wikiless.ditatompel.com/wiki/Arrow_(symbol)#Unicode) or two radio buttons could be used to show the difference between ascending/descending?

It looks like MUI library have upwards and downwards arrows for use when searching their icons:  https://mui.com/material-ui/material-icons/?query=Arrow

**Reply by ItsAllAboutTheCode** [reply]
I have prototyped the suggestions locally and attached a video of how it looks.

I placed the Background playback modal as a submenu of the "Display" menu. 
However personally I think that menu might be better of under the top level "Playback" menu when first navigating to the  Settings menu. Thoughts?

https://github.com/user-attachments/assets/31c5c41b-68bb-4d2c-97c3-6c509bf9f4d0

**Reply by erikbucik** [reply]
Yes, that's more or less what I had in mind.

> For the modal container, do you know of any reference that can be used as an example?

If you open client settings on the experimental layout, the page opens in a floating container - that's what I was thinking. [(This is the MUI component.)](https://mui.com/material-ui/react-modal/) _(And we don't need to use another obnoxious Save button.)_

> But instead have the button be on the "Display Settings" page?
> Would it be its own web page?
> However personally I think that menu might be better of under the top level "Playback" menu when first navigating to the Settings menu. Thoughts?

I'm not sure; Showing the backdrops (images) is a display setting, the rest are playback - but not really because the settings under Playback are for the user-initiated playback of media, theme media is an experience perk.

> or two radio buttons could be used to show the difference between ascending/descending?

You're right, radios will look better than the arrows here. 
Although, perhaps just a clickable button next to the "Sort by" field could be enough.

.

I think this should be good enough for now. We'll eventually redesign the whole settings part of the app but I want to update the libraries' UI first as users usually open the settings at most once.

**Reply by ItsAllAboutTheCode** [reply]
Thanks for the feedback! 
I will see if I can place the clickable button for Ordering the queue next to the Sort By dropdown.

I understand not wanting to have the background playback settings under the "Playback" menu because as you mentioned it is a passive user experience. Keeping the "Playback" as a menu for active triggering by the user makes sense.

I'll start on prototyping the changes in the experimental layout with the Modal MUI component.

**Reply by erikbucik** [reply]
Here are a couple options for the sort order dropdown UI:
1. each option duplicated (a bit much when you have many sorting options)
2. an arrow toggle button next to the dropdown
3. the order options are part of the dropdown (you need to confirm the choice before closing the dropdown)
4. an arrow before the sorting option (you have to click the option again to toggle ascending/descending order)

![sorting_dropdowns](https://github.com/user-attachments/assets/2f7f7b8e-580c-4d9f-9358-ebc47ee1ea10)
_I personally don't really like that green checkbox next to the selected option._

Which one is the best? I'm not sure. Maybe the 2nd or 4th?

**Reply by ItsAllAboutTheCode** [reply]
I have created a prototype of the UI changes for the Experimental layout which uses a [MUI Dialog](https://mui.com/material-ui/react-dialog/) component.
I went with the second approach of having an Icon Button with an arrow that sits next to the dropdown box.

I have also updated the styling of the legacy UI so that the arrow buttons fit the style a bit better. A video has been attached for both.

Please let me know what you think.

https://github.com/user-attachments/assets/40ede94e-31a3-4e80-9eb9-59a53d580e13


https://github.com/user-attachments/assets/b1644ae4-8a1e-488b-a3ba-e8db4a14c170

**Reply by ItsAllAboutTheCode** [reply]
I went ahead and created another prototype for the Background playback settings using the [MUI Popover](https://mui.com/material-ui/react-popover/) component.  The Popover component anchors the content next to the button that is pressed, so it is visually nearer to the element.

https://github.com/user-attachments/assets/2aa9e937-1826-4313-9db3-d490ac3defc1


The change can be demoed at https://6b58621a.jellyfin-web.pages.dev/

**Reply by ItsAllAboutTheCode** [reply]
Following up on this, Is there any more feedback on the UI changes for the background playback popover in experimental mode as well as the new menu in the classic UI?

**Reply by ItsAllAboutTheCode** [reply]
Pinging discussion to get more eyes and feedback on the UI changes.
